conan/tools/cmake/toolchain.py

```diff
@@ -394,29 +395,48 @@ def context(self):
         return ctxt_toolchain
 
 
-class FindConfigFiles(Block):
+class FindFiles(Block):
     template = textwrap.dedent("""
         {% if find_package_prefer_config %}
         set(CMAKE_FIND_PACKAGE_PREFER_CONFIG {{ find_package_prefer_config }})
         {% endif %}
-        # To support the generators based on find_package()
-        {% if cmake_module_path %}
-        set(CMAKE_MODULE_PATH {{ cmake_module_path }} ${CMAKE_MODULE_PATH})
+        {% if generators_folder or cmake_module_path %}
+        # To support find_package() of CMake Find files (host context), and include() of CMake modules (host & build context)
+        set(CMAKE_MODULE_PATH {{ generators_folder }} {{ cmake_module_path }} ${CMAKE_MODULE_PATH})
         {% endif %}
-        {% if cmake_prefix_path %}
-        set(CMAKE_PREFIX_PATH {{ cmake_prefix_path }} ${CMAKE_PREFIX_PATH})
+        {% if generators_folder or cmake_prefix_path %}
+        # To support find_package() of CMake config Files from host context
+        set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE "BOTH")
+        set(CMAKE_PREFIX_PATH {{ generators_folder }} {{ cmake_prefix_path }} ${CMAKE_PREFIX_PATH})
         {% endif %}
-        {% if android_prefix_path %}
-        set(CMAKE_FIND_ROOT_PATH {{ android_prefix_path }} ${CMAKE_FIND_ROOT_PATH})
+        {% if cmake_program_path %}
+        # To support find_program() of executables from build context (and host context if not cross-building)
+        set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM "BOTH")
```

Comment:
These `BOTH` values worry me a bit. That means it always will go search the system for programs, config files, libraries, and so on, right? I think this should probably be opt-in (when cross-building?) because the general rule (except cross-building, sysroot, etc scenarios) is to look only at the "conan" folders. Am I missing something?

Reply:
I can't disagree, I would prefer to not manipulate `CMAKE_FIND_ROOT_PATH_MODE_*`.

Actually, if any of these values is set to `ONLY`, the associated find command won't look into conan folders, because in this case only paths in `CMAKE_FIND_ROOT_PATH` and `CMAKE_SYSROOT` would be considered.

Unfortunately, the current conan paradigm (specifically layout of installed packaged) completly prevents to use `CMAKE_FIND_ROOT_PATH` in a robust way.

Reply:
Here what could be done for `ONLY`, and the side effects:

if `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` `ONLY`:
- add `CMAKE_CURRENT_LIST_DIR` & `builddirs` of requires (except root package folders) to `CMAKE_FIND_ROOT_PATH`.

if `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY` or `CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK` or `CMAKE_FIND_ROOT_PATH_MODE_INCLUDE` `ONLY`:
- add `builddirs` of requires to `CMAKE_FIND_ROOT_PATH`
- side effect: make executables living in requires discoverable by `find_program()`, and with the top priority, unless `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is `NEVER`.

if `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` `ONLY`:
- add `builddirs` of build requires to `CMAKE_FIND_ROOT_PATH`
- side effet: make headers, libs and frameworks in build requires discoverable by `find_file()`, `find_path()` and `find_library()`, unless `CMAKE_FIND_ROOT_PATH_MODE_<INCLUDE/LIBRARY/FRAMEWORK>` are `NEVER`.

Reply:
So, proposal to improve this PR:

- if cross-building:
  - if `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` is `ONLY`:
    - add `CMAKE_CURRENT_LIST_DIR` & `builddirs` of requires (except root package folders) to `CMAKE_FIND_ROOT_PATH`.
  - if concatanation of `bindirs` of build_requires is empty or `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is `NEVER`:
    - if `CMAKE_FIND_ROOT_PATH_MODE_INCLUDE` is `ONLY` or `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY` is `ONLY` or `CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK` is `ONLY`:
      - add `builddirs` of requires to `CMAKE_FIND_ROOT_PATH`.
  - if concatanation of `bindirs` of build_requires is not empty and `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is not `NEVER`:
    - if any of `CMAKE_FIND_ROOT_PATH_MODE_INCLUDE`, `CMAKE_FIND_ROOT_PATH_MODE_LIBRARY`, `CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK` is `ONLY`, force the variable to `BOTH`.